### PR TITLE
feat(cache): add DeleteBlobs to ArtifactCache

### PR DESCRIPTION
**Pull request by afdesk**

### CODE REVIEW

File: cache/s3.go [hunk 1]
```diff
@@ -39,6 +39,14 @@ func (c S3Cache) PutArtifact(artifactID string, artifactConfig types.ArtifactInf
 	return nil
 }
 
+func (c S3Cache) DeleteBlob(blobID string) error {
+	err := c.DeleteBlob(blobID)
```
Comment (knqyf263) [inline cache/s3.go:43]: Looks like infinite loop. Am I missing something?

Comment (afdesk) [reply cache/s3.go]: sure, my mistake. fixed

File: artifact/image/image.go [hunk 2]
```diff
@@ -120,6 +120,10 @@ func (a Artifact) Inspect(ctx context.Context) (types.ArtifactReference, error)
 	}, nil
 }
 
+func (_ Artifact) Clean(_ types.ArtifactReference) error {
```
Comment (knqyf263) [inline artifact/image/image.go:123]: nit

```suggestion
func (Artifact) Clean(_ types.ArtifactReference) error {
```

File: artifact/remote/git.go [hunk 3]
```diff
@@ -81,6 +81,10 @@ func (a Artifact) Inspect(ctx context.Context) (types.ArtifactReference, error)
 	return ref, nil
 }
 
+func (_ Artifact) Clean(_ types.ArtifactReference) error {
```
Comment (knqyf263) [inline artifact/remote/git.go:84]:
```suggestion
func (Artifact) Clean(_ types.ArtifactReference) error {
```

File: cache/cache.go [hunk 4]
```diff
@@ -28,6 +28,9 @@ type ArtifactCache interface {
 
 	// PutBlob stores blob information such as layer information in local cache
 	PutBlob(blobID string, blobInfo types.BlobInfo) (err error)
+
+	// DeleteBlob removes blob by Id
+	DeleteBlob(blobID string) error
```
Comment (knqyf263) [inline cache/cache.go:33]: After some consideration, I feel like it should take a slice.

```suggestion
	DeleteBlobs(blobIDs []string) error
```

File: cache/redis.go [hunk 5]
```diff
@@ -50,6 +50,14 @@ func (c RedisCache) PutBlob(blobID string, blobInfo types.BlobInfo) error {
 	}
 	return nil
 }
+func (c RedisCache) DeleteBlob(blobID string) error {
```
Comment (knqyf263) [inline cache/redis.go:53]: Want to add tests

File: artifact/image/image.go (hunk 2 above)
Comment (afdesk) [reply artifact/image/image.go]: done

File: artifact/remote/git.go (hunk 3 above)
Comment (afdesk) [reply artifact/remote/git.go]: done

File: cache/cache.go (hunk 4 above)
Comment (afdesk) [reply cache/cache.go]: done

File: cache/fs.go [hunk 6]
```diff
@@ -113,6 +115,21 @@ func (fs FSCache) GetArtifact(artifactID string) (types.ArtifactInfo, error) {
...
+	fs.db.Update(func(tx *bolt.Tx) error {
+		bucket := tx.Bucket([]byte(blobBucket))
+		for _, blobID := range blobIDs {
+			if err := bucket.Delete([]byte(blobID)); err != nil {
+				errs = multierror.Append(errs, err)
+			}
+		}
+		return nil
+	})
+	return errs
```
Comment (knqyf263) [inline cache/fs.go:130]: Also, we should handle an error from `Update`.

```suggestion
	err := fs.db.Update(func(tx *bolt.Tx) error {
		bucket := tx.Bucket([]byte(blobBucket))
		for _, blobID := range blobIDs {
			if err := bucket.Delete([]byte(blobID)); err != nil {
				errs = multierror.Append(errs, err)
			}
		}
		return nil
	})
	if err != nil {
		return xerrors.Errorf("DB update error: %w", err)
	}
	return errs
```

**Review by knqyf263 — CHANGES_REQUESTED** [review]

File: cache/fs.go (hunk 6 above)
Comment (afdesk) [reply cache/fs.go]: done

Comment (afdesk) [reply cache/fs.go]: oh @knqyf263 I missed.
is it really possible that `fs.db.Update` will return any error in this case?

Comment (afdesk) [reply cache/fs.go]: we accumulate all inner errors into `errs`...

Comment (knqyf263) [reply cache/fs.go]: Yes, it is possible to return an error. For example, the database is locked and db.Update can't open the database for write operations. The database is broken for some reason. etc...

File: cache/redis.go (hunk 5 above)
Comment (afdesk) [reply cache/redis.go]: added

File: cache/s3.go [hunk 7]
```diff
@@ -39,6 +39,16 @@ func (c S3Cache) PutArtifact(artifactID string, artifactConfig types.ArtifactInf
 	return nil
 }
 
+func (c S3Cache) DeleteBlob(blobID string) error {
```
[the remark this answers is not in the document]
Comment (afdesk) [reply cache/s3.go]: added tests

File: cache/fs.go (hunk 6 above)
Comment (afdesk) [reply cache/fs.go]: oh, sure. thanks for explanation!

File: cache/redis_test.go [hunk 8]
```diff
@@ -500,3 +502,62 @@ func TestRedisCache_Clear(t *testing.T) {
...
+				addr = "dummy:16379"
+			}
+
+			c := cache.NewRedisCache(&redis.Options{
+				Addr: addr,
+			})
+
+			err = c.DeleteBlobs(tt.args.blobIDs)
+			if tt.wantErr != "" {
+				require.NotNil(t, err)
```
Comment (knqyf263) [inline cache/redis_test.go:555]: nit
```suggestion
				require.Error(t, err)
```

File: cache/redis_test.go [hunk 9]
```diff
@@ -500,3 +502,62 @@ func TestRedisCache_Clear(t *testing.T) {
...
+				assert.Contains(t, err.Error(), tt.wantErr)
+				return
+			} else {
+				assert.NoError(t, err)
+			}
```
Comment (knqyf263) [inline cache/redis_test.go:560]: nit
```suggestion
			}
			assert.NoError(t, err)
```

File: artifact/local/fs.go [hunk 10]
```diff
@@ -165,3 +165,7 @@ func (a Artifact) Inspect(ctx context.Context) (types.ArtifactReference, error)
 		BlobIDs: []string{cacheKey},
 	}, nil
 }
+
+func (a Artifact) Clean(reference types.ArtifactReference) error {
+	return a.cache.DeleteBlobs([]string{reference.ID})
```
Comment (knqyf263) [inline artifact/local/fs.go:170]: It should be reference.BlobIDs technically. In filesystem scanning, `artifactID` and `blobIDs` are the same "by chance". They are usually different and the method called here is "DeleteBlobs". Passing `blobIDs` makes sense.
https://github.com/aquasecurity/fanal/blob/3519fe6e6c218e32a87b1912132d59891c171cfb/types/artifact.go#L94

Comment (afdesk) [reply artifact/local/fs.go]: updated

**Pull request merged**